File: app/src/main/java/it/chalmers/gamma/security/LoginThrottlingFilter.java
package it.chalmers.gamma.security;

import it.chalmers.gamma.app.throttling.ThrottlingService;
import jakarta.servlet.FilterChain;
import jakarta.servlet.ServletException;
import jakarta.servlet.http.HttpServletRequest;
import jakarta.servlet.http.HttpServletResponse;
import java.io.IOException;
import org.springframework.lang.NonNull;
import org.springframework.web.filter.OncePerRequestFilter;

public class LoginThrottlingFilter extends OncePerRequestFilter {

  private final ThrottlingService throttlingService;

  public LoginThrottlingFilter(ThrottlingService throttlingService) {
    this.throttlingService = throttlingService;
  }

  @Override
  protected void doFilterInternal(
      @NonNull HttpServletRequest request,
      @NonNull HttpServletResponse response,
      @NonNull FilterChain filterChain)
      throws ServletException, IOException {
    String clientIp = request.getRemoteAddr();

    if ("POST".equalsIgnoreCase(request.getMethod())
        && "/login".equals(request.getRequestURI())
        && !throttlingService.canProceed("login:" + clientIp, 50)) {
      response.sendRedirect("/login?throttle=true");
    } else {
      filterChain.doFilter(request, response);
    }
  }
}

File: app/src/main/kotlin/it/chalmers/gamma/LoginThrottlingFilter.kt
package it.chalmers.gamma

import it.chalmers.gamma.throttling.FixedWindowThrottling
import it.chalmers.gamma.throttling.ThrottleKey
import jakarta.servlet.FilterChain
import jakarta.servlet.http.HttpServletRequest
import jakarta.servlet.http.HttpServletResponse
import org.springframework.web.filter.OncePerRequestFilter
import java.time.Duration

internal class LoginThrottlingFilter(
    private val throttling: FixedWindowThrottling,
) : OncePerRequestFilter() {
    override fun doFilterInternal(
        request: HttpServletRequest,
        response: HttpServletResponse,
        filterChain: FilterChain,
    ) {
        if (request.method.equals("POST", ignoreCase = true) && request.requestURI == request.contextPath + "/login") {
            val key = ThrottleKey.digest("login", request.remoteAddr)
            val allowed = throttling.charge(key, MAXIMUM_LOGIN_ATTEMPTS, LOGIN_WINDOW)
            if (!allowed) {
                response.sendRedirect(request.contextPath + "/login?throttle=true")
                return
            }
        }
        filterChain.doFilter(request, response)
    }
}

private const val MAXIMUM_LOGIN_ATTEMPTS = 50
private val LOGIN_WINDOW: Duration = Duration.ofHours(24)
